lib/src/package_config.dart

```diff
@@ -0,0 +1,145 @@
+
+  /// Creats a package configuration with the provided avilable [packages].
+  ///
+  /// The packages must be valid packages (valid package name, valid
+  /// absolute directory URIs, valid language version, if any),
+  /// and there must not be two packages with the same name or with
+  /// overlapping root directories.
+  ///
+  /// The version of the resulting configuration is always [maxVersion].
+  factory PackageConfig(Iterable<Package> packages) =>
+      SimplePackageConfig(maxVersion, packages);
+
+  /// The configuration version number.
+  ///
+  /// Currently this is 1 or 2, where
+  /// * Version one is the `.packages` file format and
+  /// * Version two is the first `package_config.json` format.
+  ///
+  /// Instances of this class supports both, and the version
+  /// is only useful for detecting which kind of file the configuration
+  /// was read from.
+  int get version;
+
+  /// All the available packages of this configuration.
+  ///
+  /// No two of these packages have the same name,
+  /// and no two [Package.root] directories overlap.
+  Iterable<Package> get packages;
+
+  /// Look up a package by name.
+  ///
+  /// Returns the [Package] fron [packages] with [packageName] as
+  /// [Package.name]. Returns `null` if the package is not available in the
+  /// current configuration.
+  Package /*?*/ operator [](String packageName);
+
+  /// Provides the associated package for a specific [file] (or directory).
+  ///
+  /// Returns a [Package] which contains the [file]'s path.
+  /// That is, the [Package.rootUri] directory is a parent directory
+  /// of the [file]'s location.
+  /// Returns `null` if the file does not belong to any package.
+  Package /*?*/ packageOf(Uri file);
+
+  /// Resolves a `package:` URI to a non-package URI
+  ///
+  /// The [packageUri] must be a valid package URI.
+  ///
+  /// Returns the remaining path of the package URI is resolved relative to the
+  /// [Package.packageUriRoot] of the corresponding package.
+  /// Returns `null` if the package name of [packageUri] is not available
+  /// in this package configuration.
+  Uri /*?*/ resolve(Uri packageUri);
+
+  /// The package URI which resolves to [nonPackageUri].
+  ///
+  /// The [nonPackageUri] must not have any query or fragment part,
+  /// and it must not have `package` as scheme.
+  ///
+  /// Returns the a package URI which [resolve] will convert to [nonPackageUri],
+  /// if any such URI exists. Returns `null` if no such package URI exists.
+  Uri /*?*/ toPackageUri(Uri nonPackageUri);
+}
+
+/// Configuration data for a single package.
+abstract class Package {
+  /// Creates a package with the provided properties.
+  ///
+  /// The [name] must be a valid package name.
+  /// The [root] must be an absolute directory URI, meaning an absolute URI
+  /// with no query or fragment parth and a path starting and ending with `/`.
+  /// The [packageUriRoot], if provided, must be either an absolute
```

Comment:
It's probably true that in all current cases, it's either `lib/` or omitted. 
References to `.pub_cache/.../` won't need anything but the `lib/` dir, so they can make the `lib/` dir be the `rootUri` and omit `packageUriRoot`, because it then defaults to the same thing.

I think it's better to not hard-code specific file or directory names in the format.

Reply:
There is a very clear and documented package layout which requires a `lib` directory.

Given that we have a format that supports versioning, this feature seems unnecessary. Some later version can add something like this if we have a use case - and it would be a lot more likely to actually fit the requirements of said use case.

Trying to plan ahead for unknown uses adds complexity and may actually make it harder to solve for the real use cases we encounter later.

Reply:
The `lib` directory is a part of the Pub package development and distribution format, but it's not needed for package consumption. There is no requirement that the Pub cache needs to store its package libraries inside `lib` directories. Pub could have chosen to have the contents of the `lib` directory of package `foo` version 1.0.0 directly under `.pub-cache/hosted/pub.dartlang.org/foo-1.0.0/` instead of in `.pub_cache/hosted/pub/foo-1.0.0/lib/` (and saved all the space used by the non-`lib` directory files which you cannot access using a `package:` URI anyway). 
I prefer to not dictate the internal structure of implementation details.